### PR TITLE
Multiclient

### PR DESCRIPTION
**Pull request by Chenyu-Shi**
Multiclient support:
users can add certificate to the enclave, and the usernames are taken from the certificate.
Currently the demo file has two users, A and B. A and B first add their symmetric keys to the enclave. A trains a model and obtains its prediction using its own key. B then uses the same model to predict on the same testing data. B's prediction should be the same as A's.
The demo file is in demo/python/jupyter/new-demo.py.
Formatting adjustment needed.
Also, dumpmodel needs to incorporate username.

### CODE REVIEW

**Comment by Chenyu-Shi** [comment]
All the commits above are mine

File: CMakeLists.txt [hunk 1]
```diff
@@ -11,7 +11,7 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake/modules")
 cmake_policy(SET CMP0022 NEW)
 
 # OE configuration options for enclave
-set(OE_DEBUG 0
+set(OE_DEBUG 1
```
Comment (chester-leung) [inline CMakeLists.txt:14]: Please revert this file

File: demo/python/jupyter/e2e-demo.ipynb [hunk 2]
```diff
@@ -33,7 +33,7 @@
     "import securexgboost as xgb\n",
     "import os\n",
     "\n",
-    "HOME_DIR = os.path.abspath('') + \"/../../../\"\n",
+    "HOME_DIR = os.getcwd() + \"/../../../\"\n",
```
Comment (chester-leung) [inline demo/python/jupyter/e2e-demo.ipynb]: Please revert this file except for any API changes

File: enclave/ecalls.cpp [hunk 3]
```diff
@@ -36,9 +36,9 @@ int enclave_XGDMatrixCreateFromFile(const char *fname, int silent, DMatrixHandle
   return XGDMatrixCreateFromFile(fname, silent, out);
 }
 
-int enclave_XGDMatrixCreateFromEncryptedFile(const char *fname, int silent, DMatrixHandle *out) {
+int enclave_XGDMatrixCreateFromEncryptedFile(const char *fname, int silent, DMatrixHandle *out, char* username) {
   LOG(DEBUG) << "Ecall: XGDMatrixCreateFromEncryptedFile";
-  return XGDMatrixCreateFromEncryptedFile(fname, silent, out);
+  return XGDMatrixCreateFromEncryptedFile(fname, silent, out,username);
```
Comment (chester-leung) [inline enclave/ecalls.cpp]: Please add a space here between the comma and `username`

File: enclave/ecalls.cpp [hunk 4]
```diff
@@ -85,16 +85,16 @@ int enclave_XGBoosterEvalOneIter(BoosterHandle handle, int iter, DMatrixHandle d
   return XGBoosterEvalOneIter(handle, iter, dmats, (const char**) eval_names, len, (const char**) out_str);
 }
 
-int enclave_XGBoosterLoadModel(BoosterHandle handle, const char *fname) {
+int enclave_XGBoosterLoadModel(BoosterHandle handle, const char *fname, char * username) {
```
Comment (chester-leung) [inline enclave/ecalls.cpp]: We'd like to keep consistent formatting throughout our C++ code -- please change the argument in function signature here and in all other relevant places to either `char* username` or `char *username`

File: enclave/include/enclave_context.h [hunk 5]
```diff
@@ -57,11 +68,12 @@ class EnclaveContext {
...
-    bool get_client_key(uint8_t* key) {
-      if (client_key_is_set)
-          memcpy(key, client_key, CIPHER_KEY_SIZE);
-      else
-          memset(key, 0, CIPHER_KEY_SIZE);
+    bool get_client_key(uint8_t* key, char * username) {
+
+      // returning the key held by user
+        
+        LOG(INFO) << "User requested username " << username;
```
Comment (chester-leung) [inline enclave/include/enclave_context.h]: Please align the code here

File: python-package/securexgboost/core.py [hunk 6]
```diff
@@ -370,7 +370,7 @@ class DMatrix(object):
     def __init__(self, data, encrypted=False, label=None, missing=None,
                  weight=None, silent=False,
                  feature_names=None, feature_types=None,
-                 nthread=None):
+                 nthread=None, username = None):
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Please change this to `username=None`, i.e. without the spaces

File: python-package/securexgboost/core.py [hunk 7]
```diff
@@ -1502,7 +1538,7 @@ def eval(self, data, name='eval', iteration=0):
 
     def predict(self, data, output_margin=False, ntree_limit=0, pred_leaf=False,
                 pred_contribs=False, approx_contribs=False, pred_interactions=False,
-                validate_features=True):
+                validate_features=True, username = None):
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Please change this to `username=None`, i.e. without the spaces

**Review by chester-leung — CHANGES_REQUESTED** [review]
Thanks for this! I added some comments throughout your code.

Couple things:
1. Could you please delete the shell script that installs cmake? `cmake-3.15.6-Linux-x86_64.sh`

2. What's the difference between `e2e-demo.py` and `new-demo.py`? Are these Jupyter notebooks or Python scripts?

3. Could you please move all the code, keys, and certificates needed to run your example to a new folder (as opposed to adding it to `demo/python/jupyter`? Maybe you can make a new folder, `demo/python/multiclient`.  After doing this we should see no changes to any files in `demo/python/jupyter`.

4. Is your code backwards compatible? i.e. will it still be possible to add a client key and load data inside an enclave without having to specify a username? I can imagine scenarios where there may be only one lone party using our system. @podcastinator thoughts?

5.  What will happen if a file is encrypted with the current dummy test key of all zeros? Currently our `basic` and `c-api` demos assume that the files they load in are encrypted with all zeros. Basically, if the enclave sees that no client key has been added, [it will load all data assuming that the data has been encrypted with all zeros](https://github.com/mc2-project/mc2-xgboost/blob/master/enclave/include/enclave_context.h#L64). Is this still the case after your changes?

**Comment by Chenyu-Shi** [comment]
2. e2e-demo.py and new-demo.py are python scripts generated by the jupyter notebook file. new-demo.py has the latest update but i do need to strip out the jupyter comment.
3. That sounds like a good idea. I will create a new demo folder and clean up the things a bit.
4. things are definitely not backward compatible. I talked about the concept of a default key but I think @podcastinator was leaning towards specifying a username.
5. no first of all I changed the function signature second I got rid of the line that zero fills the key array.
Thanks for the review I will incorporate those changes later today

**Comment by podcastinator** [comment]
Thanks for putting this together!
I think the username should not be required with every user-facing Python API. Instead, perhaps we can have an `init` function where the user registers their username / private key etc. Then, every Python API call implicitly passes the registered username etc. to the backend code.
What do you think?

File: demo/python/jupyter/e2e-demo.ipynb (hunk 2 above)
Comment (Chenyu-Shi) [reply demo/python/jupyter/e2e-demo.ipynb]: done

File: enclave/ecalls.cpp (hunk 3 above)
Comment (Chenyu-Shi) [reply enclave/ecalls.cpp]: done

File: enclave/ecalls.cpp (hunk 4 above)
Comment (Chenyu-Shi) [reply enclave/ecalls.cpp]: done good point

**Comment by Chenyu-Shi** [comment]
So I made it such that if get_client_key defaults returns zero key like before when the username does not have an associated key, but wouldn't this be a security problem like you can just get the unencrypted prediction by providing a bogus username?

File: python-package/securexgboost/core.py (hunk 6 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: done

File: python-package/securexgboost/core.py (hunk 7 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: done

**Comment by chester-leung** [comment]
Can you elaborate on the security issue you see? What do you mean you can just get the unencrypted prediction with a bogus username? 

From my understanding, when the enclave receives a command to serve predictions, it should check that the command was issued by a user who added their key into the username => key map. If it finds the username in the map, it should then encrypt predictions with the corresponding key. I believe any set of predictions returned to a user should never be unencrypted.

**Comment by Chenyu-Shi** [comment]
I agree that should be the ideal situation. I am just concerned about what happens when the enclave can't find the username. 
This is old code.
```
bool get_client_key(uint8_t* key) {
      if (client_key_is_set)
          memcpy(key, client_key, CIPHER_KEY_SIZE);
      else
          memset(key, 0, CIPHER_KEY_SIZE);
    }
```
And this is what I have to get client key
```
    bool get_client_key(uint8_t* key, char *username) {

      // returning the key held by user
      LOG(INFO) << "User requested username " << username;
      if (client_keys.count(username) > 0){
        memcpy(key,(uint8_t *) &client_keys[CharPtrToString(username)][0], CIPHER_KEY_SIZE);
      }
      else
      {
        LOG(INFO) << "User " << username << " does not exist, using zero key instead";
        memset(key, 0, CIPHER_KEY_SIZE);
      }
    }
```
In both cases the default action is returning a zero key which is useful in c-api i belief? But this also means when you call predict on a user that does not exist, you get return value that is the unencrypted prediction.

File: python-package/securexgboost/core.py [hunk 8]
```diff
@@ -1175,6 +1190,39 @@ def add_client_key(self, data, data_len, signature, sig_len):
...
+        # Cast data : proto.NDArray to pointer to pass into C++ add_client_key()
+        data = proto_to_pointer(data)
+        data_len = ctypes.c_size_t(data_len)
+
+        # Cast signature : proto.NDArray to pointer to pass into C++ add_client_key()
+        signature = proto_to_pointer(signature)
+        sig_len = ctypes.c_size_t(sig_len)
+
+        # Add client key
+        print(_LIB.add_client_key_with_certificate(certificate, cert_len, data, data_len, signature, sig_len))
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: What are you printing here?

File: python-package/securexgboost/core.py [hunk 9]
```diff
@@ -355,6 +355,14 @@ def _maybe_dt_array(array):
 
     return array
 
+def switch_user(user_name):
```
Comment (chester-leung) [inline python-package/securexgboost/core.py]: Let's change the name of this function to `set_user`, though I see why you called it `switch_user`. In the outsourced computation model, each client has one username, and shouldn't need to switch user midway through computation. 

I think changing the name to `set_user` should alleviate any potential confusion in the outsourced computation model while maintaining clarity in your example. Thoughts?

If you change the name please remember to also change the corresponding code in `python-package/securexgboost/__init__.py`

**Review by chester-leung — CHANGES_REQUESTED** [review]
Thanks for your changes! Just a few small things.

File: CMakeLists.txt (hunk 1 above)
Comment (chester-leung) [reply CMakeLists.txt:14]: Actually let's leave it like this for now.

**Comment by chester-leung** [comment]
Also, not sure if you already have this, but should we add something in Python in each function that requires a username to error gracefully if a username is not provided to the function and not set as a global variable? Maybe an error message along the lines of "Please set your username with the `set_user` function"?

File: enclave/include/enclave_context.h (hunk 5 above)
Comment (Chenyu-Shi) [reply enclave/include/enclave_context.h]: fixed and added boolean return value

File: enclave/xgboost.edl [hunk 10]
```diff
@@ -172,6 +178,14 @@ enclave {
                 [in, size=sig_len] uint8_t* signature,
                 size_t sig_len);
 
+        public int enclave_add_client_key_with_certificate(
+                [in, size = cert_len] char * cert,
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply enclave/xgboost.edl]: Fixed

File: host/src/c_api/c_api.cc [hunk 11]
```diff
@@ -285,8 +285,9 @@ int XGDMatrixCreateFromFile(const char *fname,
 
 int XGDMatrixCreateFromEncryptedFile(const char *fname,
         int silent,
-        DMatrixHandle *out) {
-    safe_ecall(enclave_XGDMatrixCreateFromEncryptedFile(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, fname, silent, out));
+        DMatrixHandle *out,
+      char* username) {
+    safe_ecall(enclave_XGDMatrixCreateFromEncryptedFile(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, fname, silent, out,username));
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply host/src/c_api/c_api.cc]: Fixed!

File: host/src/c_api/c_api.cc [hunk 12]
```diff
@@ -919,29 +920,32 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
                              int option_mask,
                              unsigned ntree_limit,
                              xgboost::bst_ulong *len,
-                             uint8_t **out_result) {
-    safe_ecall(enclave_XGBoosterPredict(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, dmat, option_mask, ntree_limit, len, out_result));
+                             uint8_t **out_result,
+                           char* username) {
+    safe_ecall(enclave_XGBoosterPredict(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, dmat, option_mask, ntree_limit, len, out_result,username));
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply host/src/c_api/c_api.cc]: fixed

File: host/src/c_api/c_api.cc [hunk 13]
```diff
@@ -919,29 +920,32 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
...
 }
 
 
 XGB_DLL int XGBoosterLoadModelFromBuffer(BoosterHandle handle,
                                  const void* buf,
-                                 xgboost::bst_ulong len) {
-  safe_ecall(enclave_XGBoosterLoadModelFromBuffer(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, buf, len));
+                                 xgboost::bst_ulong len,
+                               char* username) {
+  safe_ecall(enclave_XGBoosterLoadModelFromBuffer(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, buf, len,username));
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply host/src/c_api/c_api.cc]: fixed! sorry that was a replace all issue.

File: host/src/c_api/c_api.cc [hunk 14]
```diff
@@ -919,29 +920,32 @@ XGB_DLL int XGBoosterPredict(BoosterHandle handle,
...
 }
 
 XGB_DLL int XGBoosterGetModelRaw(BoosterHandle handle,
                          xgboost::bst_ulong* out_len,
-                         const char** out_dptr) {
-  safe_ecall(enclave_XGBoosterGetModelRaw(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, out_len, (char**)out_dptr));
+                         const char** out_dptr,
+                       char* username) {
+  safe_ecall(enclave_XGBoosterGetModelRaw(Enclave::getInstance().getEnclave(), &Enclave::getInstance().enclave_ret, handle, out_len, (char**)out_dptr,username));
```
[the remark this answers is not in the document]
Comment (Chenyu-Shi) [reply host/src/c_api/c_api.cc]: fixed

File: python-package/securexgboost/core.py (hunk 8 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: that was in there for debugging. Just removed it

File: python-package/securexgboost/core.py (hunk 9 above)
Comment (Chenyu-Shi) [reply python-package/securexgboost/core.py]: make sense I agree. Just did a replace all.

**Comment by Chenyu-Shi** [comment]
Yeah I agree that can help our users. Just added valueerror assertions whenever a username is not provided. Thanks for the suggestion

**Review by chester-leung — APPROVED** [review]

**Comment by chester-leung** [comment]
This looks great! Thanks!

**Pull request merged**